File: day_01/day1.jl
#=
day1:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-04
=#

# file = open("day_01/test.txt")
file = open("day_01/data.txt")
data = readlines(file)

# Part 1

function count(testdata)
    highest = [0, 0] # Elf and calories
    elf::Int = 1
    total_calories::Int = 0
    for cals in testdata
        if length(cals) > 0
            calories = parse(Int, cals)
            total_calories += calories
        else
            if total_calories > highest[2]
                highest = [elf, total_calories]
            end
            elf += 1
            total_calories = 0
        end
    end
    return highest
end

highest = count(data)
println("Elf ",highest[1]," has most calories: ",highest[2])

# Part 2
# Need to keep record of the top 3 elves
function countall(testdata)
    all = []
    total_calories::Int = 0
    for cals in testdata
        if length(cals) > 0
            calories = parse(Int, cals)
            total_calories += calories
        else
            push!(all, total_calories)
            total_calories = 0
        end
    end
    return all
end

allcals = countall(data);

#=
Get the top 3, Can either use
sorted = sort(allcals, lt=(x,y)->isless(y, x));
cals = [top for top in  first(sorted ,3)]
to sort in reverse order then get the first elements, or as below, sorting in normal
order then getting the last elements.
=#

sorted = sort(allcals)
cals = [top for top in  last(sorted ,3)]
println(sum(cals))

File: day_01/day1.py
from aoc import aoc


def calories_per_elf(rows):
    calories = 0
    maxcal = 0
    for row in rows:
        if len(row) == 0:
            maxcal = max(maxcal, calories)
            calories = 0
        else:
            c = int(row)
            calories += c
    print("Calories:", maxcal)


def calories_per_elf2(rows):
    elf = 1
    allcals = aoc.CountingDict()
    for row in rows:
        if len(row) == 0:
            elf += 1
        else:
            allcals[elf] = int(row)
    return allcals


def run(file):
    print("Part 1")
    test = aoc.loadlines(file)
    calories_per_elf(test)
    print("Part 2")
    c = calories_per_elf2(test)
    sorted_keys = aoc.sort_dict(c, True)
    print(sorted_keys[:3])


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
